core/HealthCheck/HealthCheckService.php

```diff
@@ -0,0 +1,37 @@
+<?php
+
+namespace Piwik\HealthCheck;
+
+use Piwik\HealthCheck\Checks\HealthCheckInterface;
+use Piwik\HealthCheck\Responses\HealthCheckResponse;
+use Piwik\HealthCheck\Responses\HealthCheckSingleResponse;
+
+final class HealthCheckService
```

Comment:
Just a few notes from me, not necessarily required to implement but if we're adding this new API, it might be worth considering these points:

- not pass the array in constructor and provide add/remove methods that would take a healthcheck and add it to the list/remove it from the list
- the list of healthchecks could use their name as the key in the list so it can be easily removed if needed and duplicates are prevented that way if multiple plugins register the same check
- we have constants with the same name and same value in two classes (` HEALTH_CHECK_PASSED`, ` HEALTH_CHECK_FAILED`), could they be declared e.g. in this class and then used in both the single response and overall response classes?
- it seems we have too many new `final` classes here — it's just a gut feel that this might bite us in future if we want to extend things in some way. At the moment the only final class in core is the Version class. I'm not against it in principle as long as we are clear why we are using this mechanism.

Reply:
Thanks for taking a closer look!

`not pass the array in constructor and provide add/remove methods that would take a healthcheck and add it to the list/remove it from the list`

You didn't give a reason, is it for type safety? Also there is no requirement I see yet for removing.
I see the list of health checks as required, putting them in an `add` method implies they aren't really required which I think isn't the case. If it's in the constructor it's clear they are needed.

You made me take a look again and I realised it's not very defensive and could allow people to pass an empty array, this is almost definitely unintentional on their part, so I made it a little bit more defensive to make sure at least one check is passed (and added a test).

`the list of healthchecks could use their name as the key in the list so it can be easily removed if needed and duplicates are prevented that way if multiple plugins register the same check`
Good point, I don't agree with removing as there is no requirement for it, however the duplicate issue could be bothersome and cause performance issues. So I added the name as the key in the array :D, I also added a test for this as well ✅ 

`we have constants with the same name and same value in two classes ( HEALTH_CHECK_PASSED,  HEALTH_CHECK_FAILED), could they be declared e.g. in this class and then used in both the single response and overall response classes?`
Good eye! Doing that would mean I have to expose the constants as public, which implies that others can use them for something, I think for the sake of good encapsulation I will leave them where they are.


`it seems we have too many new final classes here — it's just a gut feel that this might bite us in future if we want to extend things in some way. At the moment the only final class in core is the Version class. I'm not against it in principle as long as we are clear why we are using this mechanism.`

This might be slightly a preference thing for me, but I'll explain my reasoning. Locking down the classes like this means we can avoid inheritance within this core 'package', inheritance comes with the drawbacks of tight coupling where if we change something large in the 'super class' then it breaks the others. It also means we have inflexible hierarchies if we change the super/parent class.
People can still use this behaviour if they like and do things differently with it using composition, that isn't prevented. They can still 'extend' in the sense they can easily add new health checks by implemented the interface we have.
But open to discuss this more 👍

Reply:
We can use final classes in this case I think. I can't think of any reason why this health check stuff should be extended by any plugin. The current implementation wouldn't even allow that, as plugins are not even loaded when the health check is performed.

Reply:
> > not pass the array in constructor and provide add/remove methods that would take a healthcheck and add it to the list/remove it from the list
>
> You didn't give a reason, is it for type safety? Also there is no requirement I see yet for removing.
I see the list of health checks as required, putting them in an add method implies they aren't really required which I think isn't the case. If it's in the constructor it's clear they are needed.

We probably think about it in slightly different ways but both are valid. I see the Service more as a singleton/a service provided by DI when needed, where you add/remove checks from it etc. How it's currently built is that it's expected to be instantiated every time with one or more checks. Both approaches are ok. There's no need for adding/removing checks when the whole instance can be thrown away and recreated as needed.

> You made me take a look again and I realised it's not very defensive and could allow people to pass an empty array, this is almost definitely unintentional on their part, so I made it a little bit more defensive to make sure at least one check is passed (and added a test).

👍 

> > the list of healthchecks could use their name as the key in the list so it can be easily removed if needed and duplicates are prevented that way if multiple plugins register the same check
> 
> Good point, I don't agree with removing as there is no requirement for it, however the duplicate issue could be bothersome and cause performance issues. So I added the name as the key in the array :D, I also added a test for this as well ✅

👍 

> > we have constants with the same name and same value in two classes ( HEALTH_CHECK_PASSED,  HEALTH_CHECK_FAILED), could they be declared e.g. in this class and then used in both the single response and overall response classes?
> 
> Good eye! Doing that would mean I have to expose the constants as public, which implies that others can use them for something, I think for the sake of good encapsulation I will leave them where they are.

There's slightly different approach to what's public in Matomo and to our semver commitment compared with what others might be used to, I was overly cautions at first but basically the gist is that we don't follow strict semver from PHP's perspective.
We have a list of APIs on https://developer.matomo.org/api-reference and that is our public surface (plus methods with the `@api` annotation), so having a publicly defined constant doesn't mean we can't change it. It's for a separate discussion.

In this instance here I'd say the constants and their values are actually the same/duplicates and should be reused.

> > it seems we have too many new final classes here — it's just a gut feel that this might bite us in future if we want to extend things in some way. At the moment the only final class in core is the Version class. I'm not against it in principle as long as we are clear why we are using this mechanism.
>
> This might be slightly a preference thing for me, but I'll explain my reasoning. Locking down the classes like this means we can avoid inheritance within this core 'package', inheritance comes with the drawbacks of tight coupling where if we change something large in the 'super class' then it breaks the others. It also means we have inflexible hierarchies if we change the super/parent class.
> People can still use this behaviour if they like and do things differently with it using composition, that isn't prevented. They can still 'extend' in the sense they can easily add new health checks by implemented the interface we have.
> But open to discuss this more 👍

It's all good, can leave it `final`, thanks for the explanation. Since we don't do proper DI and the service isn't implementing an interface and there are different views on what patterns we do/could/should use it's more than ok as it is :)

Reply:
```We probably think about it in slightly different ways but both are valid. I see the Service more as a singleton/a service provided by DI when needed, where you add/remove checks from it etc. How it's currently built is that it's expected to be instantiated every time with one or more checks. Both approaches are ok. There's no need for adding/removing checks when the whole instance can be thrown away and recreated as needed.
```

I also see it as this way, possibly a nice improvement for the future would be to make the list of checks configurable, so that users could specify what checks they want to perform and not have to touch core.

I believe the current approach is fine for DI still, you can always manually define the constructor params in your DI configuration.

```
In this instance here I'd say the constants and their values are actually the same/duplicates and should be reused.
```
I've extracted them to a simple enum/constant class 👍